demes/demes.py

```diff
@@ -165,27 +165,96 @@ def isclose(
         *,
         rel_tol=_ISCLOSE_REL_TOL,
         abs_tol=_ISCLOSE_ABS_TOL,
+        raise_ex=False,
     ) -> bool:
-        return (
-            self.__class__ is other.__class__
-            and isclose(
-                self.start_time, other.start_time, rel_tol=rel_tol, abs_tol=abs_tol
-            )
-            and isclose(self.end_time, other.end_time, rel_tol=rel_tol, abs_tol=abs_tol)
-            and isclose(
-                self.initial_size, other.initial_size, rel_tol=rel_tol, abs_tol=abs_tol
-            )
-            and isclose(
-                self.final_size, other.final_size, rel_tol=rel_tol, abs_tol=abs_tol
-            )
-            and self.size_function == other.size_function
-            and isclose(
-                self.selfing_rate, other.selfing_rate, rel_tol=rel_tol, abs_tol=abs_tol
-            )
-            and isclose(
-                self.cloning_rate, other.cloning_rate, rel_tol=rel_tol, abs_tol=abs_tol
-            )
-        )
+        """
+        Returns true if the epoch and ``other`` implement essentially
+        the same epoch. Compares values of the following attributes:
+        ``start_time``, ``end_time``, ``initial_size``, ``final_size``,
+        ``size_function``, ``selfing_rate``, ``cloning_rate``.
+
+        :param other: The epoch to compare against.
+        :type other: :class:`.Epoch`
+        :param ret_tol: The relative tolerance permitted for numerical
+                        comparisons. See documentation for :func:`math.isclose`
+        :type ret_tol: float
+        :param abs_tol: The absolute tolerance permitted for numerical
+                        comparisons. See documentation for :func:`math.isclose`.
+        :type abs_tol: float
+        :param raise_ex: Determines if exception (ValueError) be raised when
+                         epochs are not close. The error will contain
+                         information about attributes that are not equal.
+        :type raise_ex: bool
+
+        :return: True if the two epochs are equivalent, False otherwise.
+        :rtype: bool
+        """
+        if self.__class__ is not other.__class__:
+            if raise_ex:
```

Comment:
It might be simpler to just always raise an exception. I.e. put everything in the `isclose` methods inside a big `try` block, and check `if raises:` once when the exception is raised.